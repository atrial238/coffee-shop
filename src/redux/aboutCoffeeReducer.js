import {getSpecificCoffeeAPI} from '../API/api';

const setIsError = (bool) => ({type: types.SET_IS_ERROR, body: bool}),
		setIsLoading = (bool) => ({type: types.SET_IS_LOADING, body: bool}),
		setBestCoffee = (data) => ({type: types.GET_SPECIFIC_COFFEE, body: data}),
		setNotFound = (bool) => ({type: types.SET_NOT_FOUND, body: bool});

export const types = {
	GET_SPECIFIC_COFFEE: 'aboutCoffee_reducer/GET_BEST_COFFEE',
	SET_IS_LOADING: 'aboutCoffee_reducer/SET_IS_LOADING',
	SET_IS_ERROR: 'aboutCoffee_reducer/SET_IS_ERROR',
	SET_NOT_FOUND: 'aboutCoffee_reducer/SET_NOT_FOUND',} 

export const getSpecificCoffee = (id, methodAPI = getSpecificCoffeeAPI) => (dispatch) => {
	dispatch(setNotFound(false))
	dispatch(setIsError(false))
	dispatch(setIsLoading(true))
	return methodAPI(id)
		.then(res => {
			dispatch(setIsLoading(false))
			if(!res.specificCoffee){
				dispatch(setNotFound(true))
			}else{
				dispatch(setBestCoffee(res.specificCoffee))
			}
		})
		.catch(() => dispatch(setIsError(true)))
}

const initState = {
	specificCoffee: {},
	isLoading: false,
	isError: false,
	notFound: false
}

export const aboutCoffeeReducer = (state = initState, action) => {
	switch(action.type){
		case types.GET_SPECIFIC_COFFEE:
			return {...state, specificCoffee: action.body};
		case types.SET_IS_LOADING:
			return {...state, isLoading: action.body};
		case types.SET_IS_ERROR:
			return {...state, isError: action.body};
		case types.SET_NOT_FOUND:
			return {...state, notFound: action.body} 
		default: 
			return state;
	}
}
