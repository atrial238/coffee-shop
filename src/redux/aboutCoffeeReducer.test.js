import { getSpecificCoffee, aboutCoffeeReducer, types } from './aboutCoffeeReducer';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const mockStore = configureMockStore([thunk]);

const mockServiceCreator = (body, succeeds = true) => () =>
	new Promise((resolve, reject) => {
		setTimeout(() => (succeeds ? resolve(body) : reject('error')), 10);
});

describe('aboutCoffeeReducer', () => {
	const initState = {
		specificCoffee: {},
		isLoading: false,
		isError: false,
		notFound: false
	}

	test('shoud return initial state', () => {
		const initialState = aboutCoffeeReducer(undefined, {});
		expect(initialState).toMatchSnapshot();
	});

	test('shoud set bestCoffee array', () => {
		const action =  {type: types.GET_SPECIFIC_COFFEE, body: {coffeeone: 'one'}};
		const state = aboutCoffeeReducer(initState, action);
		expect(state).toMatchSnapshot();
	})

	test('shoud set isLoading true', () => {
		const action =  {type: types.SET_IS_LOADING, body: true};
		const state = aboutCoffeeReducer(initState, action);
		expect(state).toMatchSnapshot();
	})

	test('shoud set isError true', () => {
		const action = {type: types.SET_IS_ERROR, body: true};
		const state = aboutCoffeeReducer(initState, action);
		expect(state).toMatchSnapshot();
	})

	test('shoud set notFound true', () => {
		const action = {type: types.SET_NOT_FOUND, body: true};
		const state = aboutCoffeeReducer(initState, action);
		expect(state).toMatchSnapshot();
	})
})

describe('getBestCoffee thunk', () => {
	let store;

	beforeEach(() => store = mockStore({}));
	
	it('dispatch action to store when getSpecificCoffee success', () => {
	
	const expectedActions = [
		{type: types.SET_NOT_FOUND, body: false},
		{type: types.SET_IS_ERROR, body: false},
		{type: types.SET_IS_LOADING, body: true},
		{type: types.SET_IS_LOADING, body: false},
		{type: types.GET_SPECIFIC_COFFEE, body: 'data'}
	]
	
		return store.dispatch(getSpecificCoffee(1, mockServiceCreator({specificCoffee: 'data'})))
			.then(() => expect(store.getActions()).toEqual(expectedActions));
	});

	it('dispatch action to store when getSpecificCoffee failed', () => {
	
		const expectedActions = [
			{type: types.SET_NOT_FOUND, body: false},
			{type: types.SET_IS_ERROR, body: false},
			{type: types.SET_IS_LOADING, body: true},
			{type: types.SET_IS_ERROR, body: true},
		]
		
		return store.dispatch(getSpecificCoffee(1, mockServiceCreator(null, false)))
			.then(() => expect(store.getActions()).toEqual(expectedActions))
			.catch(() => expect(store.getActions()).toEqual(expectedActions));
	});
 })