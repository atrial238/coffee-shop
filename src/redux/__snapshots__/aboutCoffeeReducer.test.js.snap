// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aboutCoffeeReducer shoud return initial state 1`] = `
Object {
  "isError": false,
  "isLoading": false,
  "notFound": false,
  "specificCoffee": Object {},
}
`;

exports[`aboutCoffeeReducer shoud set bestCoffee array 1`] = `
Object {
  "isError": false,
  "isLoading": false,
  "notFound": false,
  "specificCoffee": Object {
    "coffeeone": "one",
  },
}
`;

exports[`aboutCoffeeReducer shoud set isError true 1`] = `
Object {
  "isError": true,
  "isLoading": false,
  "notFound": false,
  "specificCoffee": Object {},
}
`;

exports[`aboutCoffeeReducer shoud set isLoading true 1`] = `
Object {
  "isError": false,
  "isLoading": true,
  "notFound": false,
  "specificCoffee": Object {},
}
`;

exports[`aboutCoffeeReducer shoud set notFound true 1`] = `
Object {
  "isError": false,
  "isLoading": false,
  "notFound": true,
  "specificCoffee": Object {},
}
`;

exports[`homeReducer shoud return initial state 1`] = `
Object {
  "bestCoffe": Array [],
  "isError": false,
  "isLoading": false,
}
`;

exports[`homeReducer shoud set bestCoffee array 1`] = `
Object {
  "bestCoffe": Array [
    Object {
      "coffeeone": "one",
    },
    Object {
      "coffeeTwo": "two",
    },
  ],
  "isError": false,
  "isLoading": false,
}
`;

exports[`homeReducer shoud set isError true 1`] = `
Object {
  "bestCoffe": Array [],
  "isError": true,
  "isLoading": false,
}
`;

exports[`homeReducer shoud set isLoading true 1`] = `
Object {
  "bestCoffe": Array [],
  "isError": false,
  "isLoading": true,
}
`;
